Improving menu item workflow in Joomla[GSoC 16_menu_for_article_project]

shubhamnba2009 opened the pull request:
### Summary of Changes

In Joomla 1.0 it was possible to create a menu link to an article from the article edit view in the administrator back-end. Joomla is missing this feature since 1.5 and it has a long time since then. Content-Createmenulink plugin recreates the old memories of creating menu item directly from the component view.
### Testing Instructions
1. Install the patch/PR.
2. Go to Extensions > Plugins.
3. Enable Content-Createmenulink Plugin if it is disabled.
4. Click on Content-Createmenulink plugin.
5. Parameter dialogue view will appear.
   ![firstpr1](https://cloud.githubusercontent.com/assets/12000068/17907191/53158230-6998-11e6-9907-bfaede689cba.png)
6. Enter component+view in the form [com_content.article], just like in the given image below.
   Type two component view to enable component in both component view.
   [Note: Remember to press enter before typing second component+view name]

![firstpr2](https://cloud.githubusercontent.com/assets/12000068/17907253/b1230d48-6998-11e6-9e8a-71d0a310c8d7.png)
1. Navigate to content > article >create new article.
2. A new tab named as "Menu link options" will appear.
   ![firstpr3](https://cloud.githubusercontent.com/assets/12000068/17907351/2eb066d4-6999-11e6-84a7-55d1c1fea31f.png)
3. Fill all the details.
4. Check fornt end.
5. Also try to make a new menu item for contact or other 3rd party component using this PR.
   ![firstpr4](https://cloud.githubusercontent.com/assets/12000068/17907406/82d224b4-6999-11e6-95ce-89b8a3003eaf.png)
### Expected outcome

New menu item for the article/contact/3rd party component will get created in the frontend.

Commit:
Update default_enabled.php

Add changes to appear in hathor template.

Commit:
Merge remote-tracking branch 'refs/remotes/joomla-projects/staging' into staging

Commit:
Merge remote-tracking branch 'refs/remotes/joomla-projects/staging' into staging

Commit:
Merge remote-tracking branch 'refs/remotes/joomla-projects/staging' into staging

Commit:
Created xml file & language file for menu plugin

Commit:
Created xml file & language file for menu plugin

Commit:
Added new line after the end of file

Commit:
Added new line after the end of file

Commit:
Made the php file,xml file to add field.

Also updated en-GB.plg_content_menu.ini file according to xml file.

Commit:
Made the php file,xml file to add field.

Also updated en-GB.plg_content_menu.ini file according to xml file.

Commit:
Changed the menus folder name to forms and comment correction made.

Also the class name is changed and made according to the Joomla coding
standards.

Commit:
Changed the menus folder name to forms and comment correction made.

Also the class name is changed and made according to the Joomla coding
standards.

Commit:
Previous issued related to field display removed

1. Added<fieldset> tab in menu.xml so that menulink tab appears in the
backend of edit view.
2. Tested the code generic functionality with com_contact component and
it worked successfully.
3. Added a new functionality that make sure that the plugin works in the
backend of the component.

Commit:
Previous issued related to field display removed

1. Added<fieldset> tab in menu.xml so that menulink tab appears in the
backend of edit view.
2. Tested the code generic functionality with com_contact component and
it worked successfully.
3. Added a new functionality that make sure that the plugin works in the
backend of the component.

Commit:
Removed PHPCS errors

Commit:
Removed PHPCS errors

Commit:
This commit saves menu data

1. Menu data is saved in the menu table.
2. Checked with content and contact component.

Commit:
This commit saves menu data

1. Menu data is saved in the menu table.
2. Checked with content and contact component.

Commit:
removed conflicts

Commit:
removed conflicts

Commit:
Merge remote-tracking branch 'refs/remotes/joomla-projects/staging' into menuplugin

Commit:
This commit retrieves the fields of menu link tab

1. Click on already created article or contact.
2. The menu link tab fields are retrieved from the database.

Commit:
This commit retrieves the fields of menu link tab

1. Click on already created article or contact.
2. The menu link tab fields are retrieved from the database.

Commit:
Added form and clear session data

Commit:
Added form and clear session data

Commit:
Updated fucntion parameter defination

Commit:
Updated fucntion parameter defination

Commit:
Made a new field that provides option to select view of component

Coded xml file to made a drop down list of component view in which to
enable the plugin.
Fixed the decapricated, getsession() method.

Commit:
Made a new field that provides option to select view of component

Coded xml file to made a drop down list of component view in which to
enable the plugin.
Fixed the decapricated, getsession() method.

Commit:
Chaanged field name

Commit:
Chaanged field name

Commit:
Improve code style

Commit:
Improve code style

Commit:
This commit removes issues

1. Solves the problem specified in :
https://github.com/joomla-projects/gsoc16_menu-for-article/issues/16#issuecomment-235854193
2. Solves the problem specified in:
https://github.com/joomla-projects/gsoc16_menu-for-article/issues/19
3. Solves the issue specified in :

Commit:
This commit removes issues

1. Solves the problem specified in :
https://github.com/joomla-projects/gsoc16_menu-for-article/issues/16#issuecomment-235854193
2. Solves the problem specified in:
https://github.com/joomla-projects/gsoc16_menu-for-article/issues/19
3. Solves the issue specified in :

Commit:
Merge remote-tracking branch 'refs/remotes/joomla-projects/staging' into menuplugin

Commit:
Changed the plugin name to suit more perfectly

Also made necessary changes in the plugin to suit changed plugin name

Commit:
Changed the plugin name to suit more perfectly

Also made necessary changes in the plugin to suit changed plugin name

Commit:
Coded automatic suggestion of menu title based on edit view title

Commit:
Coded automatic suggestion of menu title based on edit view title

Commit:
Improved the alignment

Commit:
Improved the alignment

Commit:
Fixed the bug

Commit:
Fixed the bug

Commit:
Removed inline javascript and made seperate folder.

Improved code style

Commit:
Removed inline javascript and made seperate folder.

Improved code style

Commit:
Try to remove js bug

Commit:
Try to remove js bug

Commit:
Removed the finals bugs in making external javascript file

Commit:
Removed the finals bugs in making external javascript file

Commit:
Replaced dropdown list with text area in parameter field

Commit:
Replaced dropdown list with text area in parameter field

Commit:
Tried to use jquery to remove parent type bug

Commit:
Tried to use jquery to remove parent type bug

Commit:
Improved coding way

1. Changed the method to retrieve the view and component name of the
component.

Commit:
Improved coding way

1. Changed the method to retrieve the view and component name of the
component.

Commit:
Merge remote-tracking branch 'refs/remotes/joomla-projects/staging' into menuplugin

pe7er committed:
some edits in language file

I tried to improve the description a bit. And I would change the name "Content - Createmenulink" to "Content - Create Menu Link" to fit with Joomla's core plugins.

pe7er committed:
some edits in language file

I tried to improve the description a bit. And I would change the name "Content - Createmenulink" to "Content - Create Menu Link" to fit with Joomla's core plugins.

Commit:
Merge pull request #1 from pe7er/patch-1

some edits in language file

Commit:
Made some new changes to improve code(Read Description)

1. The issue regarding appearnce of menu link tab using explode function
is solved.
2. The issue reagrding getting view in the onContentAfterSave() is
solved.
3. Made some changes to improve code under the hood :+1:

Commit:
Made some new changes to improve code(Read Description)

1. The issue regarding appearnce of menu link tab using explode function
is solved.
2. The issue reagrding getting view in the onContentAfterSave() is
solved.
3. Made some changes to improve code under the hood :+1:

Commit:
Added code to get menuordering dynamically

1. Made a new folder named fields which contains all the required fields
specified in the xml file

Commit:
Added code to get menuordering dynamically

1. Made a new folder named fields which contains all the required fields
specified in the xml file

Commit:
Merge remote-tracking branch 'refs/remotes/joomla-projects/staging' into menuplugin

Commit:
Updated require_once with JLoader::register

Commit:
Some new improvement

Commit:
Merge branch 'menuplugin' of https://github.com/shubhamnba2009/gsoc16_menu-for-article into menuplugin

# Conflicts:
#	media/plg_content_createmenulink/parentitem.js
#	plugins/content/createmenulink/createmenulink.php
#	plugins/content/createmenulink/createmenulink.xml
#	plugins/content/createmenulink/fields/menuordering.php

Commit:
Merge remote-tracking branch 'refs/remotes/joomla/staging' into staging

Commit:
Merge remote-tracking branch 'shugsoc/menuplugin' into menupluginpr

brianteeman commented on administrator/templates/hathor/html/mod_menu/default_enabled.php:
```diff
@@ -91,7 +91,7 @@
 	{
 		$menu->addChild(
-			new JMenuNode(JText::_('MOD_MENU_COM_CONTENT_NEW_CATEGORY'), 'index.php?option=com_categories&task=category.add&extension=com_users', 'class:newarticle')
+			new JMenuNode(JText::_('MOD_MENU_COM_USERS_NOTE_NEW_CATEGORY'), 'index.php?option=com_categories&task=category.add&extension=com_users', 'class:newarticle')
```
are you sure about this change

shubhamnba2009 replied:
Oh yeah!!... 
I somehow unknowingly added this..
Will remove the this change..

brianteeman commented on administrator/language/en-GB/en-GB.plg_content_createmenulink.sys.ini:
```diff
@@ -0,0 +1,7 @@
+; Note : All ini files need to be saved as UTF-8
+
+PLG_CONTENT_CREATEMENULINK="Content - Createmenulink"
+PLG_CONTENT_CREATEMENULINK_XML_DESCRIPTION="Creates a new tab in edit form to provide a new functionality of making a menu item for the given component"
```
Lets make this more useable as it wont be new forever ;)

Howabout simply saying
"Provides the option to directly create a menu link to the item"

Note I am not saying component as its a link to the item not to the component

shubhamnba2009 replied:
Thanks for the suggestion :)

Commit:
Removed uncessary code and improved language files

zero-24 commented on plugins/content/createmenulink/createmenulink.php:
```diff
@@ -0,0 +1,222 @@
+			return true;
+		}
+
+		$app = JFactory::getApplication();
```
Why do you declare app here and just use it once? Maybe we should remove this var?

shubhamnba2009 replied:
You mean to directly use it as:
`if (!(JFactory::getApplication()->isAdmin()))
        {
            return true;
        }`

zero-24 replied:
Yes or use the var on all places in that method :)

zero-24 commented on plugins/content/createmenulink/createmenulink.php:
```diff
@@ -0,0 +1,222 @@
+			return true;
+		}
+
+		$app = JFactory::getApplication();
```
Same here. Maybe we should use chaining here?

zero-24 replied:
same here maybe remove var and use chaingin

zero-24 replied:
same here with the app var :)

zero-24 commented on plugins/content/createmenulink/fields/menuordering.php:
```diff
@@ -0,0 +1,108 @@
+		$query = $db->getQuery(true)
+			->select('a.id AS value, a.title AS text')
+			->from('#__menu AS a')
+
```
Is there a reason for this clean line?

shubhamnba2009 replied:
No, Will make the correction

zero-24 commented:
As we add a plugin we need The Update SQLs, add the new plugin to the install sql, the new plugin needs to be added to the script.php and the new lanugage files needs to be added to the install.xml

zero-24 commented on plugins/content/createmenulink/createmenulink.php:
```diff
@@ -0,0 +1,222 @@
+	 *
+	 * @since   3.6
+	 */
+
```
Here is a clean line to much :)

shubhamnba2009 replied:
Noted :) :+1:

andrepereiradasilva commented on plugins/content/createmenulink/createmenulink.php:
```diff
@@ -0,0 +1,222 @@
+/**
+ * Createmenulink Plugin
+ *
+ * @since  3.6
```
please use `@since  __DEPLOY_VERSION__` in all since comments

andrepereiradasilva commented on plugins/content/createmenulink/createmenulink.xml:
```diff
@@ -0,0 +1,37 @@
+    <license>GNU General Public License version 2 or later; see LICENSE.txt</license>
+    <authorEmail>admin@joomla.org</authorEmail>
+    <authorUrl>www.joomla.org</authorUrl>
+    <version>1.1.0</version>
```
`__DEPLOY_VERSION__`

andrepereiradasilva commented on plugins/content/createmenulink/createmenulink.xml:
```diff
@@ -0,0 +1,37 @@
+        <filename plugin="createmenulink">createmenulink.php</filename>
+	    <folder>fields</folder>
+	    <folder>forms</folder>
+    </files>
```
please use tabs, not spaces to indent

zero-24 commented:
We should move `media/plg_content_createmenulink/copytitle.js` to `media/plg_content_createmenulink/js/copytitle.js` and add a `copytitle.min.js` same for the other JS file.

zero-24 commented on plugins/content/createmenulink/createmenulink.php:
```diff
@@ -0,0 +1,222 @@
+	public function onContentBeforeSave($context, $article)
+	{
+		$session = JFactory::getSession();
+		$session->set("formData", JFactory::getApplication()->input->post->get('jform', array(), 'array'));
```
we can use here:

```
JFactory::getSession()->set("formData", JFactory::getApplication()->input->post->get('jform', array(), 'array'));
```

zero-24 commented on plugins/content/createmenulink/fields/menuparent.php:
```diff
@@ -0,0 +1,89 @@
+		// Merge any additional options in the XML definition.
+		$options = array_merge(parent::getOptions(), $options);
+
+		return $options;
```
any reason to not do the following?

```
// Merge any additional options in the XML definition.
return array_merge(parent::getOptions(), $options);
```

shubhamnba2009 replied:
Done in this commit: https://github.com/joomla/joomla-cms/pull/11766/commits/98c0cfefd7e3e635ba3b4e834d0d1398703994fe

zero-24 committed:
Just a bit CS on te XML form

andrepereiradasilva commented on administrator/language/en-GB/en-GB.plg_content_createmenulink.ini:
```diff
@@ -0,0 +1,24 @@
+PLG_CONTENT_CREATEMENULINK_MENUS_ITEM_FIELD_ORDERING_VALUE_FIRST="- First -"
+PLG_CONTENT_CREATEMENULINK_MENUS_ITEM_FIELD_ORDERING_VALUE_LAST="- Last -"
+PLG_CONTENT_CREATEMENULINK_MENUS_ITEM_FIELD_PARENT_LABEL="Parent Item"
+PLG_CONTENT_CREATEMENULINK_MENUS_ITEM_FIELD_PARENT_DESC="Select a parent item."
```
please alpha order all language strings

shubhamnba2009 replied:
Done in this commit: https://github.com/joomla/joomla-cms/pull/11766/commits/98c0cfefd7e3e635ba3b4e834d0d1398703994fe

zero-24 commented:
@shubhamnba2009 please also see: https://github.com/shubhamnba2009/joomla-cms/pull/1

zero-24 commented on plugins/content/createmenulink/createmenulink.php:
```diff
@@ -0,0 +1,222 @@
+
+use Joomla\Registry\Registry;
+
+JLoader::import('joomla.application.component.model');
```
@mbabker can you double check that? I think we don't need this as it got autoloaded or I'm wrong here?

mbabker replied:
Not needed.  This isn't even importing a valid file.

zero-24 replied:
Thanks so we should remove this line :smile: @shubhamnba2009

shubhamnba2009 replied:
Removed in the latest commit: https://github.com/joomla/joomla-cms/pull/11766/commits/98c0cfefd7e3e635ba3b4e834d0d1398703994fe :)

zero-24 commented on plugins/content/createmenulink/createmenulink.php:
```diff
@@ -0,0 +1,222 @@
+		// Check we are manipulating a valid form.
+		$name = $form->getName();
+
+		$checkView = preg_split("/\\r\\n|\\r|\\n/", $this->params->get('allowedContexts'));
```
maybe we should move that $checkView lines to a private method? so we just need to call it and not copy the code?

zero-24 commented on plugins/content/createmenulink/createmenulink.php:
```diff
@@ -0,0 +1,222 @@
+		}
+
+		$session = JFactory::getSession();
+		$data    = $session->get("formData");
```
I personaly think it looks better with using `'` insted of `"` in that get and clear etc. can you fix that in the complete file?

zero-24 committed:
Update createmenulink.xml

zero-24 commented:
and https://github.com/shubhamnba2009/joomla-cms/pull/2 :smile:

jeckodevelopment committed:
Copyright year update

jeckodevelopment commented:
see also minor change in shubhamnba2009#3

jeckodevelopment committed:
Copyright year update

andrepereiradasilva commented on plugins/content/createmenulink/createmenulink.php:
```diff
@@ -0,0 +1,222 @@
+	 * @var    boolean
+	 * @since  3.6
+	 */
+	protected $autoloadLanguage = true;
```
is there a reason to autoload the language file in ALL joomla frontend and backend pages?

Commit:
Merge pull request #1 from zero-24/patch-16

Just a bit CS on te XML form

Commit:
Merge pull request #3 from jeckodevelopment/patch-1

Copyright year update in the js files

Commit:
Merge pull request #2 from zero-24/patch-18

Fix up the createmenulink.xml

Commit:
Merge remote-tracking branch 'refs/remotes/joomla/staging' into menupluginpr

Commit:
Fixed some issues

1. Replaced version no. version with __DEPLOY_VERSION__
2. Made some CS changes.
3. Replaced " with ' quotes.
4. Made changes in the language file to make it in the alphabetical
order
5. Removed extra spaces from code.

andrepereiradasilva commented:
@shubhamnba2009 as @zero-24 said

> As we add a plugin we need The Update SQLs, add the new plugin to the install sql, the new plugin needs to be added to the script.php and the new lanugage files needs to be added to the install.xml

You need to add the plugin to the installation sql (mysql: https://github.com/joomla/joomla-cms/blob/staging/installation/sql/mysql/joomla.sql#L616) and also to the update sql (mysql: https://github.com/joomla/joomla-cms/tree/staging/administrator/components/com_admin/sql/updates/mysql) in all 3 database systems.

Also you need to add them to updater so the manifest cache can be refreshed on update: https://github.com/joomla/joomla-cms/blob/staging/administrator/components/com_admin/script.php#L348

brianteeman commented on administrator/language/en-GB/en-GB.plg_content_createmenulink.sys.ini:
```diff
@@ -0,0 +1,7 @@
+; License GNU General Public License version 2 or later; see LICENSE.txt, see LICENSE.php
+; Note : All ini files need to be saved as UTF-8
+
+PLG_CONTENT_CREATEMENULINK="Content - Createmenulink"
```
For consistency please change this to 
"PLG_CONTENT_CREATEMENULINK="Content - Create Menu Link"

brianteeman commented on administrator/language/en-GB/en-GB.plg_content_createmenulink.ini:
```diff
@@ -0,0 +1,24 @@
+PLG_CONTENT_CREATEMENULINK_COMPONENT_OPTION_DESC="Enter the components for which you want to trigger this plugin, and the view to which it should redirect in the front-end. Use One per line in the format com_componentname.viewname"
+PLG_CONTENT_CREATEMENULINK_ITEM_FIELD_TITLE_DESC="The title of the menu item that will display in the menu."
+PLG_CONTENT_CREATEMENULINK_ITEM_FIELD_TITLE_LABEL="Menu Title"
+PLG_CONTENT_CREATEMENULINK_LINK_OPTIONS="Menu link options."
```
To follow the style guide please change this to
"Menu Link Options"

brianteeman commented on administrator/language/en-GB/en-GB.plg_content_createmenulink.ini:
```diff
@@ -0,0 +1,24 @@
+PLG_CONTENT_CREATEMENULINK_MENUS_MENUID="Menu ID"
+PLG_CONTENT_CREATEMENULINK_MENUS_MENUID_DESC="This field will save menu ID in hidden mode"
+PLG_CONTENT_CREATEMENULINK_MENUS_MENUNAME_DESC="This will create a Menu Item for this Item in the Menu that you select"
+PLG_CONTENT_CREATEMENULINK_MENUS_MENUNAME_LABEL="Menu name"
```
To follow the style guide please change this to
"Menu Name"

brianteeman commented on administrator/language/en-GB/en-GB.plg_content_createmenulink.ini:
```diff
@@ -0,0 +1,24 @@
+PLG_CONTENT_CREATEMENULINK_MENUS_ITEM_FIELD_PARENT_DESC="Select a parent item."
+PLG_CONTENT_CREATEMENULINK_MENUS_ITEM_FIELD_PARENT_LABEL="Parent Item"
+PLG_CONTENT_CREATEMENULINK_MENUS_ITEM_ROOT="Menu Item Root"
+PLG_CONTENT_CREATEMENULINK_XML_DESCRIPTION="Provides the option to directly create a menu link to the item"
```
Please add a . at the end

brianteeman commented on administrator/language/en-GB/en-GB.plg_content_createmenulink.ini:
```diff
@@ -0,0 +1,24 @@
+PLG_CONTENT_CREATEMENULINK_LINK_OPTIONS="Menu link options."
+PLG_CONTENT_CREATEMENULINK_MENUS_MENUID="Menu ID"
+PLG_CONTENT_CREATEMENULINK_MENUS_MENUID_DESC="This field will save menu ID in hidden mode"
+PLG_CONTENT_CREATEMENULINK_MENUS_MENUNAME_DESC="This will create a Menu Item for this Item in the Menu that you select"
```
Please add a . at the end

brianteeman commented on administrator/language/en-GB/en-GB.plg_content_createmenulink.ini:
```diff
@@ -0,0 +1,24 @@
+PLG_CONTENT_CREATEMENULINK_ITEM_FIELD_TITLE_LABEL="Menu Title"
+PLG_CONTENT_CREATEMENULINK_LINK_OPTIONS="Menu link options."
+PLG_CONTENT_CREATEMENULINK_MENUS_MENUID="Menu ID"
+PLG_CONTENT_CREATEMENULINK_MENUS_MENUID_DESC="This field will save menu ID in hidden mode"
```
Please add a . at the end

brianteeman commented on administrator/language/en-GB/en-GB.plg_content_createmenulink.sys.ini:
```diff
@@ -0,0 +1,7 @@
+; Note : All ini files need to be saved as UTF-8
+
+PLG_CONTENT_CREATEMENULINK="Content - Createmenulink"
+PLG_CONTENT_CREATEMENULINK_XML_DESCRIPTION="Provides the option to directly create a menu link to the item"
```
Please add a . at the end

andrepereiradasilva commented on plugins/content/createmenulink/createmenulink.php:
```diff
@@ -0,0 +1,214 @@
+		}
+
+		$jinput    = JFactory::getApplication()->input;
+		$component = $jinput->getcmd('option');
```
`getCmd`, not `getcmd` notice the Case

andrepereiradasilva commented on plugins/content/createmenulink/createmenulink.php:
```diff
@@ -0,0 +1,214 @@
+		$session->clear('formData');
+
+		$jinput    = JFactory::getApplication()->input;
+		$component = $jinput->getcmd('option');
```
`getCmd`, not `getcmd` notice the Case

andrepereiradasilva commented on plugins/content/createmenulink/forms/createmenulink.xml:
```diff
@@ -0,0 +1,62 @@
+			default="0"
+			label="PLG_CONTENT_CREATEMENULINK_MENUS_MENUID"
+			description="PLG_CONTENT_CREATEMENULINK_MENUS_MENUID_DESC"
+			filter="integer"
```
the order is:
name="_"
type="_"
label="_"
description="_"
[others]

see http://joomla.github.io/coding-standards/?coding-standards/chapters/xml.md

andrepereiradasilva commented on plugins/content/createmenulink/fields/menuordering.php:
```diff
@@ -0,0 +1,107 @@
+			$query->where('a.menutype != ' . $db->quote(''));
+		}
+
+		$query->order('a.lft ASC');
```
please quote all database fields, example:

``` php
$db->quoteName('lft')
$db->quoteName('#__menu')
```

and you don't need the `a.` alias since you are not making any join with other table...

andrepereiradasilva commented on plugins/content/createmenulink/fields/menuparent.php:
```diff
@@ -0,0 +1,87 @@
+
+		$query->where('a.published != -2')
+			->group('a.id, a.title, a.level, a.lft, a.rgt, a.menutype, a.parent_id, a.published')
+			->order('a.lft ASC');
```
please quote all database fields, example:

``` php
$db->quoteName('lft')
$db->quoteName('a.lft')
$db->quoteName('a.id', 'value')
$db->quoteName('#__menu')
$db->quoteName('#__menu', 'b')
```

andrepereiradasilva replied:
@shubhamnba2009 This is required now

> #### decide whether or not we are going to enforce the use of quoteName() and quote() in database queries. (Roland)
> 
> The PLT has decided that it will enforce the use of quoteName() and quote() methods on any new pull requests involving database queries. This to have consistent database queries that will work best across SQL platforms and with maximum security over time.

See https://volunteers.joomla.org/leadership/production-leadership-team/reports/296-plt-meeting-july-26-2016

shubhamnba2009 replied:
@andrepereiradasilva Yeah It's next on my list and I am working on it... I am understanding how quoteName() and quote() works with database queries... First time use with quoteName() and quote() :P

andrepereiradasilva replied:
exmaple

``` php
$query = $db->getQuery(true)
    ->select('a.id AS value, a.title AS text, a.level')
    ->from('#__menu AS a')
    ->join('LEFT', $db->quoteName('#__menu') . ' AS b ON a.lft > b.lft AND a.rgt < b.rgt');
```

is this

``` php
$query = $db->getQuery(true)
    ->select($db->quoteName('a.id', 'value'))
    ->select($db->quoteName('a.title', 'text'))
    ->select($db->quoteName('level'))
    ->from($db->quoteName('#__menu', 'a'))
    ->join('LEFT', $db->quoteName('#__menu', 'b') . ' ON '. $db->qn('a.lft') . ' > ' . $db->qn('b.lft') . ' AND ' . $db->qn('a.rgt') . ' < ' . $db->qn('b.rgt'));
```

note: i used the shortnote `qn` (instead of `quoteName` to keep it in one line)

shubhamnba2009 replied:
Is it wrong to go like this
`query = $db->getQuery(true)`
      `->select($db->quoteName(array('id', 'value'), array('title', 'text'),array('level)))`

andrepereiradasilva replied:
never used like that when using alias, if it works, it's fine

shubhamnba2009 replied:
`->join('LEFT', $db->quoteName('#__menu', 'b') . ' ON '. $db->qn('a.lft') . ' > ' . $db->qn('b.lft') . ' AND ' . $db->qn('a.rgt') . ' < ' . $db->qn('b.rgt'));`

It is still giving PHPCS error of using more than 150 character in a line

Commit:
Changes made from suggestion(for info read description)

1.  Added the plugin to joomla.sql in installation folder, added the new
plugin to the install (mysql,postgresql,sqlazure), script.php and the
new lanugage files added to the install.xml
2. Made grammatical changes in the language files.
3. Moved .js in the new folder named js inside media folder.
4. Added new file with extension .min.js along with .js file.
5. Made some code changes to improve code style.

andrepereiradasilva commented on administrator/components/com_admin/sql/updates/postgresql/3.7.0-2016-09-01.sql:
```diff
@@ -0,0 +1,2 @@
+INSERT INTO `#__extensions` (`extension_id`, `name`, `type`, `element`, `folder`, `client_id`, `enabled`, `access`, `protected`, `manifest_cache`, `params`, `custom_data`, `system_data`, `checked_out`, `checked_out_time`, `ordering`, `state`) VALUES
+(458, 'plg_content_createmenulink', 'plugin', 'createmenulink', 'content', 0, 1, 1, 0, '', '', '', '', 0, '0000-00-00 00:00:00', 0, 0);
```
in postgresql the dates are: 1970-01-01 00:00:00

andrepereiradasilva commented on administrator/components/com_admin/sql/updates/sqlazure/3.7.0-2016-09-01.sql:
```diff
@@ -0,0 +1,6 @@
+SET IDENTITY_INSERT [#__extensions] ON;
+
+INSERT INTO [#__extensions] ([extension_id], [name], [type], [element], [folder], [client_id], [enabled], [access], [protected], [manifest_cache], [params], [custom_data], [system_data], [checked_out], [checked_out_time], [ordering], [state])
+SELECT 458, 'plg_content_createmenulink', 'plugin', 'createmenulink', 'content', 0, 1, 1, 0, '', '', '', '', 0, '0000-00-00 00:00:00', 0, 0;
```
in sqlazure the dates are 1900-01-01 00:00:00

andrepereiradasilva commented on installation/sql/mysql/joomla.sql:
```diff
@@ -614,6 +614,7 @@ INSERT INTO `#__extensions` (`extension_id`, `name`, `type`, `element`, `folder`
 (455, 'plg_installer_packageinstaller', 'plugin', 'packageinstaller', 'installer', 0, 1, 1, 1, '', '', '', '', 0, '0000-00-00 00:00:00', 1, 0),
 (456, 'plg_installer_folderinstaller', 'plugin', 'folderinstaller', 'installer', 0, 1, 1, 1, '', '', '', '', 0, '0000-00-00 00:00:00', 2, 0),
 (457, 'plg_installer_urlinstaller', 'plugin', 'urlinstaller', 'installer', 0, 1, 1, 1, '', '', '', '', 0, '0000-00-00 00:00:00', 3, 0),
+(458, 'plg_content_createmenulink', 'plugin', 'createmenulink', 'content', 0, 1, 1, 0, '', '', '', '', 0, '0000-00-00 00:00:00', 0, 0),
```
you also need to add it to postgresql and sqlazur joomla.sql install sql file

shubhamnba2009 replied:
Thanks for the suggetsion :)
Will make the change

zero-24 commented:
We need the mssql and postgressql for thebinstall SQL too :)

Commit:
Merge remote-tracking branch 'refs/remotes/joomla/staging' into menupluginpr

Commit:
Changed date in postgresql,sqlazure and added plugin to installation folder

andrepereiradasilva commented on installation/sql/postgresql/joomla.sql:
```diff
@@ -613,7 +613,8 @@ INSERT INTO "#__extensions" ("extension_id", "name", "type", "element", "folder"
 (455, 'plg_installer_packageinstaller', 'plugin', 'packageinstaller', 'installer', 0, 1, 1, 1, '', '', '', '', 0, '1970-01-01 00:00:00', 1, 0),
 (456, 'plg_installer_folderinstaller', 'plugin', 'folderinstaller', 'installer', 0, 1, 1, 1, '', '', '', '', 0, '1970-01-01 00:00:00', 2, 0),
-(457, 'plg_installer_urlinstaller', 'plugin', 'urlinstaller', 'installer', 0, 1, 1, 1, '', '', '', '', 0, '1970-01-01 00:00:00', 3, 0);
+(457, 'plg_installer_urlinstaller', 'plugin', 'urlinstaller', 'installer', 0, 1, 1, 1, '', '', '', '', 0, '1970-01-01 00:00:00', 3, 0)
```
missing a comma here ...

Commit:
Fixed missing comma in postgresql\joomla.sql file

andrepereiradasilva commented on plugins/content/createmenulink/fields/menuordering.php:
```diff
@@ -0,0 +1,107 @@
+		// Merge any additional options in the XML definition.
+		$options = array_merge(parent::getOptions(), $options);
+
+		return $options;
```
you could replace this with just

``` php
// Merge any additional options in the XML definition.
return array_merge(parent::getOptions(), $options);
```

andrepereiradasilva commented on plugins/content/createmenulink/createmenulink.xml:
```diff
@@ -0,0 +1,37 @@
+					type="textarea"
+					default="com_content.article"
+					label="PLG_CONTENT_CREATEMENULINK_COMPONENT_OPTION"
+					description="PLG_CONTENT_CREATEMENULINK_COMPONENT_OPTION_DESC"
```
the order is name, type, label, description, others

Commit:
Improved CS and removed unneccessary code

Commit:
Changed queries with quoteName() and quote()

shubhamnba2009 commented:
@andrepereiradasilva Have a look in the latest commit of changing queries with quoteName() and quote(), thanks :)

andrepereiradasilva commented on plugins/content/createmenulink/fields/menuordering.php:
```diff
@@ -0,0 +1,106 @@
+			->select($db->quoteName('a.title', 'text'))
+			->from($db->quoteName('#__menu', 'a'))
+			->where($db->quoteName('published') . ' >= ' . $db->quote('0'))
+			->where($db->quoteName('a.parent_id') . ' = ' . $db->quote((int) $parent_id));
```
please don't quote integers

also i still don't understand why you need to use the `a` table alias here. you are not making any join ...

andrepereiradasilva commented on plugins/content/createmenulink/fields/menuparent.php:
```diff
@@ -0,0 +1,96 @@
+				->where('NOT(' . $db->quoteName('a.lft') . ' >= ' . $db->qn('p.lft') . ' AND ' . $db->qn('a.rgt') . ' <= ' . $db->qn('p.rgt') . ' ) ');
+		}
+
+		$query->where($db->quoteName('a.published') . ' != ' . $db->quoteName('-2'))
```
please don't quote integers

andrepereiradasilva commented on plugins/content/createmenulink/fields/menuparent.php:
```diff
@@ -0,0 +1,96 @@
+			->group($db->quoteName('a.rgt'))
+			->group($db->quoteName('a.menutype'))
+			->group($db->quoteName('a.parent_id'))
+			->group($db->quoteName('a.published'))
```
you could use 

``` php
->group($db->quoteName(array('a.id', 'a.title', ...)))
```

Commit:
Made changes in the queries in the menuparent.php and menuordering.php

Commit:
Merge remote-tracking branch 'refs/remotes/joomla/staging' into menupluginpr

andrepereiradasilva commented on plugins/content/createmenulink/fields/menuordering.php:
```diff
@@ -0,0 +1,106 @@
+			->select($db->quoteName('title', 'text'))
+			->from($db->quoteName('#__menu'))
+			->where($db->quoteName('published') . ' >= ' . $db->quote('0'))
+			->where($db->quoteName('parent_id') . ' = (int) $parent_id ');
```
this is not correct

shubhamnba2009 replied:
@andrepereiradasilva Can you also suggest the correct syntax for this line of code?

zero-24 replied:
The parent_id is a PHP variable but you passed it as string ;)

It should be quotet too ;) the casting too (int) don't work too ;)

andrepereiradasilva replied:
``` php
->where($db->quoteName('published') . ' >= 0')
->where($db->quoteName('parent_id') . ' = ' . (int) $parent_id);
```

infograf768 commented:
As we use `menu item` and not `menu link` generally speaking in Joomla, I suggest to modify the ini string values to fit.
Also we do not use `Menu Name` but simply `Menu` (see mod_menu)

shubhamnba2009 commented:
@infograf768 That means I have to change the plugin name also?

infograf768 commented:
I would, but I am not the only one to decide. Let's wait what others think.

andrepereiradasilva commented:
o agrre with @infograf768 IIRC there is no "mwnu link" in joomla, it's always called "menu item".
for consistency all should changed to menu item IMHO.

Commit:
Changed folder name to createmenuitem and also replaced the same at some code position

Commit:
Merge remote-tracking branch 'refs/remotes/joomla/staging' into menupluginpr

Commit:
Made further changes to change name from createmenulink to createmenuitem

andrepereiradasilva commented on plugins/content/createmenuitem/forms/createmenuitem.xml:
```diff
@@ -0,0 +1,62 @@
+<?xml version="1.0" encoding="utf-8"?>
+<form>
+	<fieldset name="menulink" label="PLG_CONTENT_CREATEMENUITEM_LINK_OPTIONS" addfieldpath="/plugins/content/createmenuitem/fields">
```
menulink ?

andrepereiradasilva commented on media/plg_content_createmenulink/js/copytitle.js:
```diff
@@ -0,0 +1,24 @@
+/**
+ * @package         Joomla.JavaScript
+ * @copyright       Copyright (C) 2005 - 2016 Open Source Matters, Inc. All rights reserved.
```
all this 4 js should be in `plg_content_createmenuitem` folder not `plg_content_createmenulink`

shubhamnba2009 replied:
Thanx for double checking @andrepereiradasilva .
There were so many changes. I thought that i might have left some changes.

andrepereiradasilva replied:
sure, no problem. but you should test your plugin properly after all this changes.

andrepereiradasilva commented on administrator/language/en-GB/en-GB.plg_content_createmenuitem.ini:
```diff
@@ -0,0 +1,24 @@
+PLG_CONTENT_CREATEMENUITEM_MENUS_ITEM_FIELD_PARENT_DESC="Select a parent item."
+PLG_CONTENT_CREATEMENUITEM_MENUS_ITEM_FIELD_PARENT_LABEL="Parent Item"
+PLG_CONTENT_CREATEMENUITEM_MENUS_ITEM_ROOT="Menu Item Root"
+PLG_CONTENT_CREATEMENUITEM_XML_DESCRIPTION="Provides the option to make a menu item."
```
please alpha order the language strings
example:

```
PLG_CONTENT_CREATEMENUITEM_MENUS_MENUNAME_LABEL="Menu"
PLG_CONTENT_CREATEMENUITEM_MENUS_ITEM_FIELD_ORDERING_DESC="The menu item will be placed in the menu after the selected menu item."
```

is not alpha ordered

andrepereiradasilva commented on media/plg_content_createmenulink/js/parentitem.js:
```diff
@@ -0,0 +1,33 @@
+					$('#jform_parent_id').append(option);
+				});
+				$('#jform_parent_id').trigger('chosen:updated');
+			});
```
in ajax request you must check the session token and need to catch ajax request fail.

See https://github.com/joomla/joomla-cms/blob/staging/media/system/js/sendtestmail-uncompressed.js#L30 as an example.

Commit:
Made furtehr changes(Read below for more info)

1. Changed folder name plg_content_createmenulink to
plg_content_menuitem in media folder
2. Ordered the language string inside language file in alphabatical
order.
3. Also changed menulink to menuitem where i  missed before

Commit:
Merge remote-tracking branch 'refs/remotes/joomla/staging' into menupluginpr

dgt41 commented:
@shubhamnba2009 can you see my approach on the module ordering https://github.com/joomla/joomla-cms/pull/11040? Maybe it worth a port here to menuorder field...
Just an idea

andrepereiradasilva commented on plugins/content/createmenuitem/createmenuitem.xml:
```diff
@@ -0,0 +1,37 @@
+<?xml version="1.0" encoding="utf-8"?>
+<extension version="3.6" type="plugin" group="content" method="upgrade">
```
3.6 ?

shubhamnba2009 replied:
Replace it with **DEPLOY_VERSION** ??

andrepereiradasilva replied:
Actually not sure the `version=""` there accepts 3 digits (x.y.z) versions that `__DEPLOY_VERSION__`will add.

Commit:
Removed bugs

1. Changed folder path
2. Changed wrong query

andrepereiradasilva commented on plugins/content/createmenuitem/createmenuitem.php:
```diff
@@ -0,0 +1,214 @@
+
+		else
+		{
+			JHtml::_('script', 'media/plg_content_createmenuitem/js/copytitle.js');
```
use relative paths. example

``` php
JHtml::_('script', 'plg_content_createmenuitem/copytitle.js', false, true);
```

shubhamnba2009 replied:
But copytitle.js is inside js folder

andrepereiradasilva replied:
i know but that's the way joomla does when using relative paths
try and you see it works

Commit:
Changed the path to the relative path

Commit:
Merge remote-tracking branch 'refs/remotes/joomla/staging' into menupluginpr

Commit:
Merge remote-tracking branch 'refs/remotes/joomla/staging' into menupluginpr

shubhamnba2009 commented:
Is dis read for commit? Any more suggestions or changes that I need to implement?

brianteeman commented:
I couldnt get it to install using patchtester so I downloaded your branch to try.

that shows the plugin as being enabled but there is nothing in the article to create a menu

brianteeman commented:
I had to open and save the plugin - that suggests that you are not setting the plugin params in th installation

shubhamnba2009 commented:
@brianteeman 
Yeah I also just checked out that after opening and saving the plugin starts to works. I have updated the installation folder before. Let me have a double check on what I left. Thanks for the heads up!!

brianteeman commented:
**Issue 1**
1. Create an article called brian and save it with a menu item of brian
2. Create an article called john and save it with a menu item of brian
Expected result is that the article will NOT save and you will get an error that the menu exists already
Actual result is that the article is saved and you will get an error that the menu exists already 

**Issue 2**
1. Create an article called simon and save and close
2. Re-open the article and save and close
Expected result is that the article is saved as normal
Actual result is that that the article is saved AND a menu item is created 

Thats just the start. Every time I open an article I see something new. This needs a lot more work in testing what happens when you create articles AND edit existing articles ESPECIALLY if you do NOT want a menu item

shubhamnba2009 commented:
For issue 2.
I had considered that user had enabled plugin with the intention to create a menu item. So if does not want to get a menu item created he had to clear the menu title, which in fact has another fucntionality of getting auto-filled based on the article title. 
So basically every time user enter the article title same title will be automatically copied to menu title. Beacuse of this functionality menu item is getting created.
So what's you suggestion. Should I remove auto-fill functionality?

brianteeman commented:
YES the auto-fill **cannot** be on for everything - no way you would ever want to create a menu item for everything

brianteeman commented:
> In Joomla 1.0 it was possible to create a menu link to an article from the article edit view in the administrator back-end.

It should work like that - you have the **option/ability** to create a menu link from the article edit view. It should not be created automatically for everything

pe7er commented:
I like the auto-refill function (automatically copying the (article) title for the menu item) because it speeds up your workflow as you've to type less. However I did not realize that as @brianteeman describes, because the menu title was already added by auto-fill, it would always create a menu item.

Maybe we can change the feature a bit? e.g. add a checkbox or button with "copy menu item title from title" and if it's clicked then copy the title of the item to the Menu Item Title input box.

brianteeman commented:
It needs to be more than that. Right now the code will always create a menu
item if there is a title. There is no positive action "create menu". There
needs to be. Its more than just how the title is created that needs looking
at it is the whole process of how a menu item is created

On 10 September 2016 at 07:54, Peter Martin <notifications@github.com>
wrote:

> I like the auto-refill function (automatically copying the (article) title
> for the menu item) because it speeds up your workflow as you've to type
> less. However I did not realize that as @brianteeman
> <https://github.com/brianteeman> describes, because the menu title was
> already added by auto-fill, it would always create a menu item.
>
> Maybe we can change the feature a bit? e.g. add a checkbox or button with
> "copy menu item title from title" and if it's clicked then copy the title
> of the item to the Menu Item Title input box.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/joomla/joomla-cms/pull/11766#issuecomment-246095345>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABPH8ch49GOmZ9aX-E8bwfrJb-posxKuks5qolQ0gaJpZM4JrUoO>
> .
>



-- 
Brian Teeman
Co-founder Joomla! and OpenSourceMatters Inc.
http://brian.teeman.net/

Commit:
Removed autofill menu title feature and included duplicate menu item error

If somehow menu duplication occurs while creating a component item, now
an error will be generated to create menu item with different name.

Commit:
Merge remote-tracking branch 'refs/remotes/joomla/staging' into menupluginpr

Commit:
Merge remote-tracking branch 'refs/remotes/joomla/staging' into staging

Commit:
Update parameter field in the sql files

shubhamnba2009 commented:
@brianteeman I have done modifications as suggested by you. Could you please review once?

puneet0191 commented:
@shubhamnba2009  Thanks :+1:

Commit:
Merge remote-tracking branch 'refs/remotes/joomla/staging' into menupluginpr

jeckodevelopment commented:
FYI: PLT voted on merging this and motion passed.

@shubhamnba2009 and @puneet0191 can you please solve conflicts with this PR?

Please test again since this will be merged in the core.

Commit:
Merge remote-tracking branch 'refs/remotes/joomla/staging' into staging

Commit:
removed merge conflicts

# Conflicts:
#	administrator/components/com_admin/script.php
#	installation/sql/mysql/joomla.sql
#	installation/sql/postgresql/joomla.sql
#	installation/sql/sqlazure/joomla.sql

andrepereiradasilva commented on plugins/content/createmenuitem/createmenuitem.php:
```diff
@@ -0,0 +1,229 @@
+			$data->parent_id    = $menuItems[0]->parent_id;
+			$data->menuordering = $menuItems[0]->id;
+
+			JHtml::_('script', 'plg_content_createmenuitem/parentitem.js', false, true);
```
please use the new method signature

``` php
JHtml::_('script', 'plg_content_createmenuitem/parentitem.js', array('version' => 'auto', 'relative' => true));
```

shubhamnba2009 replied:
Thanks @andrepereiradasilva 
Can you also tell me that where I can know about all other new changes and new methods signature recently made?
I mean is there some kind of news update or I have to go through documentation?

Commit:
Merge remote-tracking branch 'refs/remotes/joomla/staging' into menupluginpr

Commit:
Added back the conflicts changes and also new method signature

# administrator/components/com_admin/script.php
# installation/sql/mysql/joomla.sql
# installation/sql/postgresql/joomla.sql
# installation/sql/sqlazure/joomla.sql

Commit:
Merge remote-tracking branch 'refs/remotes/joomla/staging' into menupluginpr

brianteeman commented:
When creating a menu item there is an option to select the parent item
I would expect this to show me a list of all the current menu items for the selected menu
Instead it shows me a list of all the menu items for all the menus and the admin menus in one single list

![menulink](https://cloud.githubusercontent.com/assets/1296369/20137821/27823cc2-a67d-11e6-9066-126dc87ddcda.gif)

infograf768 commented:
Agree with Brian.
The confusion also comes from the fact that there are 2 dropdown fields
One general when editing the Content: Parent Item
Another one in the Menu Item Options. 
Until the article is saved AFTER choosing the Menu in the Menu Item Optiins, we then get the correct possible Parent Items, and only then.

The Menu has to be chosen first chosen and the possible parent menu items should be displayed on the fly.
In any case, we should never display back-end menu items imho.

Another issue is that after save and close, if we re-edit the article, there is no trace of the menu item that has been created. This is prone to errors when one has many article menu items.

brianteeman commented:
Not just a confusion - it is 100% wrong to show those admin component menu items

infograf768 commented:
It is very easy to target only frontend menu items with a
`->where($db->quoteName('a.client_id') . ' = 0')`

But, for me, even after correcting this, the way it works is very confusing as parent menu items corresponding to a specific menu are not proposed on the fly and no trace remains of an article menu item already created for that article when editing the article.

Commit:
Update from joomla/staging

Commit:
Merge remote-tracking branch 'refs/remotes/joomla/staging' into menupluginpr

# Conflicts:
#	installation/language/sr-YU/sr-YU.ini

Commit:
update from Joomla/staging

Commit:
Merge remote-tracking branch 'refs/remotes/joomla/staging' into menupluginpr

Commit:
Merge remote-tracking branch 'refs/remotes/joomla/staging' into menupluginpr

# Conflicts:
# installation/sql/mysql/joomla.sql

Commit:
Merge remote-tracking branch 'refs/remotes/joomla/staging' into staging

Commit:
Merge remote-tracking branch 'refs/remotes/joomla/staging' into menupluginpr

Commit:
Made changes in installation\sql folder

# In mysql\joomla.sql file added the missing data for a row
# In sqlazure\jooma.sql file update createmenuitem id number

Commit:
Merge remote-tracking branch 'refs/remotes/joomla/staging' into menupluginpr

Commit:
Merge remote-tracking branch 'refs/remotes/joomla/staging' into menupluginpr

Commit:
Made changes to remove duplicate menu bugs

Commit:
Improved code style to make travis happy

Commit:
Merge branch 'staging' into menupluginpr

Commit:
Merge branch 'staging' into menupluginpr

Commit:
Merge remote-tracking branch 'refs/remotes/joomla/staging' into menupluginpr

# Conflicts:
#	administrator/components/com_admin/script.php
#	installation/sql/mysql/joomla.sql
#

Commit:
Merge remote-tracking branch 'refs/remotes/joomla/staging' into menupluginpr

# Conflicts:
#	installation/language/sr-YU/sr-YU.ini
#	installation/sql/sqlazure/joomla.sql

Commit:
Merge branch 'menupluginpr' of https://github.com/shubhamnba2009/joomla-cms into menupluginpr

Commit:
Merge remote-tracking branch 'refs/remotes/joomla/staging' into staging

# Conflicts:
#	administrator/templates/hathor/html/mod_menu/default_enabled.php

puneet0191 commented:
@shubhamnba2009 Could you please resolve conflicts on this one?

Commit:
Merge remote-tracking branch 'refs/remotes/joomla/staging' into menupluginpr

Commit:
Merge remote-tracking branch 'refs/remotes/joomla/staging' into menupluginpr

# Conflicts:
#	administrator/components/com_admin/script.php
#	installation/sql/mysql/joomla.sql
#	installation/sql/postgresql/joomla.sql
#	installation/sql/sqlazure/joomla.sql

shubhamnba2009 commented:
@puneet0191 I have noticed a new check (appveyor) is integrated now. The PR is failing its check... Any idea whom should I contact to know about it?

Commit:
Merge remote-tracking branch 'refs/remotes/joomla/staging' into menupluginpr

Commit:
Merge remote-tracking branch 'refs/remotes/joomla/staging' into menupluginpr

Commit:
removed conflicts

Commit:
Merge remote-tracking branch 'refs/remotes/joomla/staging' into menupluginpr

shubhamnba2009 committed:
Merge branch 'staging' into menupluginpr

Commit:
Merge remote-tracking branch 'refs/remotes/joomla/staging' into menupluginpr

Commit:
Merge branch 'menupluginpr' of https://github.com/shubhamnba2009/joomla-cms into menupluginpr

roland-d commented:
@shubhamnba2009 Appveyor should be fine. If you can resolve the conflicts we can see what the automated tests do and we can follow up as needed.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/11766">issues.joomla.org/tracker/joomla-cms/11766</a>.</sub>

Commit:
Merge remote-tracking branch 'refs/remotes/joomla/staging' into staging

Commit:
Merge branch 'staging' of https://github.com/joomla/joomla-cms into staging

Commit:
Merge remote-tracking branch 'refs/remotes/origin/staging' into menupluginpr

# Conflicts:
#	administrator/components/com_admin/script.php
#	administrator/language/en-GB/install.xml
#	administrator/templates/hathor/html/mod_menu/default_enabled.php
#	installation/sql/mysql/joomla.sql
#	installation/sql/postgresql/joomla.sql
#	installation/sql/sqlazure/joomla.sql

shubhamnba2009 commented:
@roland-d I resolved the conflicts.. Let me know how the automated test does :)

roland-d commented:
@shubhamnba2009 All tests passed. Thanks 👍

roland-d commented:
Now we just need human test. I will get some people to test next week.

brianteeman commented:
Tested this and it never populates the parent item

![menu](https://user-images.githubusercontent.com/1296369/29488755-543cc23a-8509-11e7-93d6-b035548176da.gif)

brianteeman commented:
Further test 
1. Menu title not stored in the article so no way to see if an article has a menu item
2. Menu item created as unpublished so you still need to go to the menu to publish it which defeats part of the object of this PR

![menu2](https://user-images.githubusercontent.com/1296369/29488783-df8d9d1e-8509-11e7-943a-28026546622c.gif)

wilsonge commented:
This has been superceeded by #21357 and merged so closing here - thanks for all the hard work that was done here!
